### PR TITLE
TYP: remove ignore from pandas/tseries/frequencies.py II

### PR DESCRIPTION
**Pull request by natmokval**
Related to pr #52623, which doesn't work correctly.
mypy ignore[assignment] was removed from pandas/tseries/frequencies.py

### CODE REVIEW

File: pandas/tseries/frequencies.py [hunk 1]
```diff
@@ -172,15 +173,14 @@ def infer_freq(
         inferer = _TimedeltaFrequencyInferer(index)
         return inferer.get_freq()
 
-    if isinstance(index, Index) and not isinstance(index, DatetimeIndex):
+    if not hasattr(index, "dtype"):
+        pass
+    elif isinstance(index.dtype, object):
```
Comment (jbrockmendel) [inline pandas/tseries/frequencies.py:178]: index.dtype is always an object

File: pandas/tseries/frequencies.py [hunk 2]
```diff
@@ -172,15 +176,12 @@ def infer_freq(
...
+    elif isinstance(index, Index):
         if is_numeric_dtype(index.dtype):
             raise TypeError(
                 f"cannot infer freq from a non-convertible index of dtype {index.dtype}"
             )
-        # error: Incompatible types in assignment (expression has type
-        # "Union[ExtensionArray, ndarray[Any, Any]]", variable has type
-        # "Union[DatetimeIndex, TimedeltaIndex, Series, DatetimeLikeArrayMixin]")
-        index = index._values  # type: ignore[assignment]
+        index = cast("TimedeltaArray", index._values)
```
Comment (jbrockmendel) [inline pandas/tseries/frequencies.py:184]: this cant be right.  TimedeltaIndex goes through the _TimedeltaFrequencyInferer path above

Comment (natmokval) [reply pandas/tseries/frequencies.py]: Thank you for your comment. If I understand correctly, we only need to check` is_numeric_dtype(index.dtype) `then.

Comment (jbrockmendel) [reply pandas/tseries/frequencies.py]: yep, looks like that should work.

Comment (natmokval) [reply pandas/tseries/frequencies.py]: thank you for reviewing this pr.

**Comment by jbrockmendel** [comment]
thanks @natmokval

**Pull request merged**